src/hotspot/share/opto/connode.cpp

```diff
@@ -46,6 +46,7 @@ uint ConNode::hash() const {
 ConNode *ConNode::make(const Type *t) {
   switch( t->basic_type() ) {
   case T_INT:         return new ConINode( t->is_int() );
+  case T_SHORT:       return new ConHNode( t->is_half_float_constant() );
```

Comment:
JVM treats, byte and short as constrained integer type, which is why we create ConI and not ConB or ConS.
In addition, transform routines of PhaseGVN and PhaseIterGVN use ConNode::make interface to create a constant IR node, it will not be appropriate to add a specialization over there. I have modified the check to remove unnecessary ambiguity while still maintaining the constant creation interface.